Source/swiftlint/Helpers/LintableFilesVisitor.swift

```diff
@@ -162,25 +163,64 @@ struct LintableFilesVisitor {
         return nil
     }
 
-    private static func loadCompileCommands(_ path: String) -> [String: [String]]? {
-        guard let jsonContents = FileManager.default.contents(atPath: path),
-            let object = try? JSONSerialization.jsonObject(with: jsonContents),
+    private static func loadCompileCommands(_ path: String) -> Result<[File: Arguments], CompileCommandsLoadError> {
+        guard let jsonContents = FileManager.default.contents(atPath: path) else {
+            return .failure(.nonExistantFile(path))
+        }
+
+        guard let object = try? JSONSerialization.jsonObject(with: jsonContents),
             let compileDB = object as? [[String: Any]] else {
-            return nil
+            return .failure(.malformedCommands(path))
         }
 
         // Convert the compilation database to a dictionary, with source files as keys and compiler arguments as values.
         //
         // Compilation databases are an array of dictionaries. Each dict has "file" and "arguments" keys.
-        return compileDB.reduce(into: [:]) { (commands: inout [File: Arguments], entry: [String: Any]) in
-            if let file = entry["file"] as? String, var arguments = entry["arguments"] as? [String] {
-                // Compilation databases include the compiler, but it's left out when sending to SourceKit.
-                if arguments.first == "swiftc" {
-                    arguments.removeFirst()
-                }
-
-                commands[file] = CompilerArgumentsExtractor.filterCompilerArguments(arguments)
+        var commands = [File: Arguments]()
+        for (index, entry) in compileDB.enumerated() {
+            guard let file = entry["file"] as? String else {
+                return .failure(.malformedFile(path, index))
+            }
+
+            guard var arguments = entry["arguments"] as? [String] else {
+                return .failure(.malformedArguments(path, index))
+            }
+
+            // Compilation databases include the compiler, but it's left out when sending to SourceKit.
+            if arguments.first == "swiftc" {
+                arguments.removeFirst()
             }
+
+            if !arguments.contains(file) {
+                return .failure(.missingFileInArguments(path, index, arguments))
+            }
```

Comment:
This is a new check, I'm a bit worried about the performance implications of this, especially since it's a small sanity check, but I added it here for discussion

Reply:
Pretty minor, but I'd prefer if this check were a `guard` above the `swiftc` removal for consistency with the rest of the error handling checks.

As for the performance overhead, I think this is trivial and completely dwarfed by the cost of running analyzer rules.